### ISSUE TITLE
Routing error - entrance and barrier ignored in favor of area center.

### ISSUE THREAD

**Issue by contrapunctus-1** [osmandapp/Osmand#7340]
OsmAnd v3.4.4 from F-Droid, on a Nokia 6.1, Android v9.

Using offline maps and in-app pedestrian routing in Delhi, India with default settings
from geo:28.569485,77.23886
to geo:28.56841,77.23811
results in the following -

![Screenshot_20190728-174038](https://user-images.githubusercontent.com/8513118/62006941-2a3d8980-b165-11e9-9197-cdc7dc9374a4.png)

The gate with entrance=main is ignored, even though there is a path connecting it to the main street and the path continues into the school. Even if a gate isn't connected (as such details may often be missing), an entrance=main should be preferred over the area center, especially if the area has a barrier=* tag.

On a side note, what on earth is up with the routing in the northeast near the station?

**Comment by hakuchi** [comment]
this is because the footways and steps near your starting point have no connection to the street.
Because of this, the engine use the nearest street to calculate the path.

check out this footway: https://www.openstreetmap.org/way/697784572
the endpoint of it is connected to the building, but not to the street: https://www.openstreetmap.org/node/6553107235

**Comment by contrapunctus-1** [comment]
That is correct - the way you've linked is a Metro interchange walkway, which is not directly connected to the streets. It might be because the station exit isn't connected to the streets, but changing the starting point to the street near the exit doesn't fix the routing issue, either :thinking:

**Comment by hakuchi** [comment]
if the way, I linked above, have no connection to a other way with `highway` tagging, you will not find any router what is able to calculate a correct route.

You can use https://wiki.openstreetmap.org/wiki/Tag:highway%3Dcorridor to add a connection from the subway to the street.

This is not a error in OsmAnd. You can fix this in OSM!

**Comment by contrapunctus-1** [comment]
@hakuchi , it isn't related to the subway not being connected to the street. Even if I move the start point to the street, the problem persists -

![Screenshot_20190925-193804](https://user-images.githubusercontent.com/8513118/66131714-8ca58100-e611-11e9-986c-f41cb9f2fbec.png)

**Comment by hakuchi** [comment]
if you will have the correct routing to the entrance of the scool you must place the endpoit near the entranse and not somewhere in the backyard.

**Comment by contrapunctus-1** [comment]
Ah, I just tapped the school, that's the end point OsmAnd sets for you. At any rate, since the boundary is a wall, I'd request that OsmAnd take that into consideration and look for an entrance=*.
